File: src/36. Valid Sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */

export function isValidSudoku(board) {
  const row = Array.from(Array(9), () => ({})),
    col = Array.from(Array(9), () => ({})),
    box = Array.from(Array(9), () => ({}))
  for (let i = 0; i < 9; i++) {
    for (let j = 0; j < 9; j++) {
      let val = board[i][j]
      if (val === '.') continue
      const k = Math.floor(i / 3) * 3 + Math.floor(j / 3)
      if (row[i][val] || col[j][val] || box[k][val]) {
        return false
      } else {
        row[i][val] = col[j][val] = box[k][val] = true
      }
    }
  }
  return true
}

File: src/36. Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        bool rows[9][9] = {false}, cols[9][9] = {false}, boxes[9][9] = {false};
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[i].size(); ++j) {
                if (board[i][j] != '.') {
                    int value = board[i][j] - '0' - 1,
                        k = i / 3 * 3 + j / 3;
                    if (rows[i][value] || cols[j][value] || boxes[k][value]) {
                        return false;
                    }
                    rows[i][value] = cols[j][value] = boxes[k][value] = true;
                }
            }
        }
        return true;
    }
};